webhooks-extension/endpoints/webhook.go

```diff
@@ -126,12 +148,142 @@ func (r Resource) createWebhook(request *restful.Request, response *restful.Resp
+				logging.Log.Errorf("error deleting the webhook (name %s), error: %s", name, err)
+				RespondError(response, err, http.StatusInternalServerError)
+			}
+		} else {
+			logging.Log.Info("Deleted the webhook OK, deleting from ConfigMap next")
+		}
+
+		err = r.deleteWebhookFromConfigMapByName(name, namespace)
+		if err != nil {
+			logging.Log.Errorf("error deleting the webhook information (name %s) from the ConfigMap, error: %s", name, err)
+			RespondError(response, err, http.StatusInternalServerError)
+		}
+	} else {
+		logging.Log.Error("no name was provided")
+		RespondError(response, err, http.StatusBadRequest)
+	}
+	response.WriteHeader(201)
+}
+
+// Find the repository URL for a webhook. We store webhook in a ConfigMap
+func (r Resource) findRepoURLFromConfigMap(webhookName, namespace string) string {
+	foundURL := ""
+	configMapClient := r.K8sClient.CoreV1().ConfigMaps(namespace)
+
+	configMap, err := configMapClient.Get(ConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		logging.Log.Errorf("couldn't find ConfigMap %s in namespace %s", ConfigMapName, namespace)
+		return ""
+	}
+
+	data := configMap.BinaryData["GitHubSource"]
+
+	itsANestedMap := map[string]map[string]string{}
+	err = json.Unmarshal(data, &itsANestedMap)
+	if err != nil {
+		logging.Log.Errorf("invalid data format in the ConfigMap, can't modify it, error: %s", err.Error())
+	}
+
+	foundURL = itsANestedMap[webhookName]["gitrepositoryurl"]
+	return foundURL
+}
+
+// Delete the webhook information from our ConfigMap
+func (r Resource) deleteWebhookFromConfigMapByName(webhookName, namespace string) error {
+	logging.Log.Debug("Deleting webhook info from ConfigMap")
+
+	configMapClient := r.K8sClient.CoreV1().ConfigMaps(namespace)
+
+	configMap, err := configMapClient.Get(ConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		logging.Log.Errorf("couldn't find the ConfigMap %s in namespace %s", ConfigMapName, namespace)
+		return err
+	} else {
+		logging.Log.Debugf("Found and deleting webhook info %s from %s", webhookName, configMap.Name)
+	}
+
+	/* Contains for example, base64 encoded:
+	{
+		confiMap.BinaryData[GitHubSource]
+		"myhook":
+		{
+			"name":"myhook",
+			"namespace":"default",
+			"gitrepositoryurl":"https://myrepourl",
+		  "accesstoken":"my-secret-for-hooks",
+		  "pipeline":"simple-helm-pipeline-insecure",
+			"dockerregistry":"adamroberts"
+		}
+	}
+	*/
+
+	data := configMap.BinaryData["GitHubSource"]
+
+	itsANestedMap := map[string]map[string]string{}
+	err = json.Unmarshal(data, &itsANestedMap)
+	if err != nil {
+		logging.Log.Errorf("invalid data format in the ConfigMap, can't modify it, error: %s", err.Error())
+	}
+
+	delete(itsANestedMap, webhookName)
+
+	itsANestedMapAsBytes, err := json.Marshal(itsANestedMap)
+	if err != nil {
+		logging.Log.Errorf("Couldn't go back to bytes from the nested map, error is: %s", err.Error())
+	}
+
+	configMap.BinaryData["GitHubSource"] = itsANestedMapAsBytes
+
+	_, err = configMapClient.Update(configMap)
```

Comment:
discussion: does there need to be some locking around updates to the config map?

Reply:
I agree, added a test case, need to fix implementation next.

If two users are attempting to update the configmap at the same time, we may end up with the wrong data (let's say Adam does a delete to remove his webhook as Duane does the same thing to remove theirs, and we're behaving particularly slow). Duane's update is going to work on the old ConfigMap's data.

https://golang.org/src/net/http/client.go the http client itself that we use mentions being thread safe...

I've added a simple test case using a WaitGroup with two goroutines, both doing requests to delete different webhooks, and then we check they were really both deleted OK (that is, one didn't act upon some invalid data). This should simulate two HTTP requests interacting with the same resource. I thought all was well but on wrapping said test in a for loop and doing it 100 times, there's a 1/10 chance (observed so far!) that sometimes not all of the data is removed successfully. So I'll add a lock.